primo/utils/tests/test_clustering_utils.py

```diff
@@ -28,117 +31,140 @@
                     "Latitude": 40.0,
                     "Longitude": -71,
                     "Age [Years]": 20,
                     "Depth [ft]": 1000,
+                    "Op Name": "Owner 1",
                 },
                 {
+                    "Well API": "W2",
                     "Latitude": 41.0,
                     "Longitude": -72,
                     "Age [Years]": 30,
                     "Depth [ft]": 2000,
+                    "Op Name": "Owner 2",
                 },
             ],
             {"distance": 0.5, "depth": 0.2, "ages": 0.3, "age": 0.3},  # Weight
-            "Currently, the feature ages is not supported in the clustering step",  # Result
+            (
+                "Received feature(s) [ages] that are not "
+                "supported in the clustering step."
+            ),  # Result
             False,  # Status
         ),
         # Add more test cases as needed
     ],
 )
 def test_distance_matrix(well_data, weight, result, status):
     well_df = pd.DataFrame(well_data)
+    well_cn = WellDataColumnNames(
+        well_id="Well API",
+        latitude="Latitude",
+        longitude="Longitude",
+        age="Age [Years]",
+        depth="Depth [ft]",
+        operator_name="Op Name",
+    )
+    wd = WellData(well_df, well_cn)
     result_arr = np.array(result)
     if status:
         assert np.allclose(
-            distance_matrix(well_df, weight), result_arr, rtol=1e-5, atol=1e-8
+            distance_matrix(wd, weight), result_arr, rtol=1e-5, atol=1e-8
         )
     else:
         with pytest.raises(ValueError):
-            _ = distance_matrix(well_df, weight) == result
+            _ = distance_matrix(wd, weight) == result
+
+
+def test_perform_clustering(caplog):
+    filename = os.path.dirname(os.path.realpath(__file__))[:-12]  # Primo folder
+    filename += "//data_parser//tests//random_well_data.csv"
+
+    col_names = WellDataColumnNames(
+        well_id="API Well Number",
+        latitude="x",
+        longitude="y",
+        operator_name="Operator Name",
+        age="Age [Years]",
+        depth="Depth [ft]",
+    )
+
+    wd = WellData(filename=filename, column_names=col_names)
+    assert "Clusters" not in wd
+    assert not hasattr(col_names, "cluster")
+
+    num_clusters = perform_clustering(wd)
+    assert "Clusters" in wd
+    assert hasattr(col_names, "cluster")
+    assert num_clusters == 18
+    assert num_clusters == len(set(wd.data["Clusters"]))
+
+    assert (
+        "Found cluster attribute in the WellDataColumnNames object."
+        "Assuming that the data is already clustered. If the corresponding "
+        "column does not correspond to clustering information, please use a "
+        "different name for the attribute cluster while instantiating the "
+        "WellDataColumnNames object."
+    ) not in caplog.text
 
+    # Capture the warning if the data has already been clustered
+    num_clusters = perform_clustering(wd)
```

Comment:
Why are we clustering the data again? (Line 146)

Reply:
Specifically to trigger the test where we check that an already clustered model is not clustered again. This is OK!